### DISCUSSION TITLE
Why can a video at the bottom of the playlist be instantly sorted?

### DISCUSSION THREAD

**Discussion by alvinchitena** [KohGeek/SortYoutubePlaylistByDuration#8] [Q&A] [answered]
If you move a video to the bottom of the video (Move to bottom), then Sort, the video is sorted in one move to the correct index. But if you press Sort without moving it to the bottom, it moves one by one to the bottom. Is there a reason for the differentiated behavior? @KohGeek

**Comment by KohGeek** [comment]
For the video that you moved to the bottom, is that the longest/shortest or random video? I forgot what algorithm the sort uses but it's either insertion sort or bubble sort, not the most efficient one to be fair

If I remember correctly, it's ~~insertion~~ selection sort, it will look at the shortest video available after the sorted n list, and move it to n+1

---

Edit: I've seen this behaviour before. Normally what happens is that the the list was sorted from 1 -> n, then n+1 (let's call it x) is larger than the values of n+2 -> n+m, and then n+2 -> n+m is sorted also. This means the sort algorithm is going to very slowly move from position of n+2 to n+1, and then position of n+3 to n+2 and so forth of until x is at n+m.

If x was placed at the end of the playlist, it will be moved to the correct n+m location in one move.

I haven't had time to develop or look into an algorithm on how to minimise the amount of API write, but it has been on my todo list for a while, if there any suggestion, feel free to drop them.

**Reply by alvinchitena** [reply]
Any random video added at the top of an already sorted playlist. Normal sort will move it down one index at a time until it reaches the correct spot. But if you move to bottom first, It is sorted to the correct index in one move.

I'm guessing it has to do with the sorting algorithm and how you are checking the state. You are sorting the first unsorted item you find. So, if it's [1, 2, 3, 4, 6, 7, 5], the first unsorted item is the last one in this case, so it can be sorted in one move. But if it's [5, 1, 2, 3, 4, 6, 7], you have to sort 1, 2, 3, 4. I see it now. Thanks for explaining.

**Reply by KohGeek** [reply, marked as answer]
Just to add, I've discovered cycle sort after very brief googling that would probably help improve sorting speed massively, I'll consider it for implementation soon, but no promises

**Reply by alvinchitena** [reply]
Thank you.
